src/keychain.js

```diff
@@ -175,223 +172,195 @@ class Keychain {
+      return throwDelayed(errcode(new Error(`Invalid key name '${name}'`), 'ERR_INVALID_KEY_NAME'))
     }
     const dsname = DsName(name)
-    self.findKeyByName(name, (err, keyinfo) => {
-      if (err) return _error(callback, err)
-      const batch = self.store.batch()
-      batch.delete(dsname)
-      batch.delete(DsInfoName(name))
-      batch.commit((err) => {
-        if (err) return _error(callback, err)
-        callback(null, keyinfo)
-      })
-    })
+    const keyInfo = await self.findKeyByName(name)
+    const batch = self.store.batch()
+    batch.delete(dsname)
+    batch.delete(DsInfoName(name))
+    await batch.commit()
+    return keyInfo
   }
 
   /**
    * Rename a key
    *
    * @param {string} oldName - The old local key name; must already exist.
    * @param {string} newName - The new local key name; must not already exist.
-   * @param {function(Error, KeyInfo)} callback
    * @returns {undefined}
    */
-  renameKey (oldName, newName, callback) {
+  async renameKey (oldName, newName) {
     const self = this
     if (!validateKeyName(oldName) || oldName === 'self') {
-      return _error(callback, errcode(new Error(`Invalid old key name '${oldName}'`), 'ERR_OLD_KEY_NAME_INVALID'))
+      return throwDelayed(errcode(new Error(`Invalid old key name '${oldName}'`), 'ERR_OLD_KEY_NAME_INVALID'))
     }
     if (!validateKeyName(newName) || newName === 'self') {
-      return _error(callback, errcode(new Error(`Invalid new key name '${newName}'`), 'ERR_NEW_KEY_NAME_INVALID'))
+      return throwDelayed(errcode(new Error(`Invalid new key name '${newName}'`), 'ERR_NEW_KEY_NAME_INVALID'))
     }
     const oldDsname = DsName(oldName)
     const newDsname = DsName(newName)
     const oldInfoName = DsInfoName(oldName)
     const newInfoName = DsInfoName(newName)
-    this.store.get(oldDsname, (err, res) => {
-      if (err) {
-        return _error(callback, errcode(new Error(`Key '${oldName}' does not exist. ${err.message}`), 'ERR_KEY_NOT_FOUND'))
-      }
-      const pem = res.toString()
-      self.store.has(newDsname, (err, exists) => {
-        if (err) return _error(callback, err)
-        if (exists) return _error(callback, errcode(new Error(`Key '${newName}' already exists`), 'ERR_KEY_ALREADY_EXISTS'))
-
-        self.store.get(oldInfoName, (err, res) => {
-          if (err) return _error(callback, err)
-
-          const keyInfo = JSON.parse(res.toString())
-          keyInfo.name = newName
-          const batch = self.store.batch()
-          batch.put(newDsname, pem)
-          batch.put(newInfoName, JSON.stringify(keyInfo))
-          batch.delete(oldDsname)
-          batch.delete(oldInfoName)
-          batch.commit((err) => {
-            if (err) return _error(callback, err)
-            callback(null, keyInfo)
-          })
-        })
-      })
-    })
+    let res = await this.store.get(oldDsname)
+    const pem = res.toString()
+
+    const exists = await self.store.has(newDsname)
```

Comment:
What about doing the new validating first, then the get of the old one?

Reply:
Makes sense, done!